core/sql/optimizer/ScmCostMethod.cpp

```diff
@@ -3632,6 +3632,561 @@ CostMethodMergeUnion::scmComputeOperatorCostInternal(RelExpr *op,
+  else
+  {
+    //  singleSubsetPrefixOrder==0  means either there
+    // is an equal, an IN,  or there are no key preds in the
+    // first column.
+    // singleSubsetPrefixOrder==0 AND itIsSingleSubset
+    // means there is an EQUAL or there are no key preds
+    // in the first column, check for existance of
+    // predicates in this case:
+    if (     itIsSingleSubset // this FALSE for an IN predicate
+	 AND keyPredsByCol[0] != NULL
+       )
+    {
+      thereAreSingleSubsetPreds = TRUE;
+    }
+  }
+
+
+  CMPASSERT(NOT histograms.isEmpty());
+
+  if ( thereAreSingleSubsetPreds )
+  {
+    // ---------------------------------------------------------
+    // There are some key predicates, so apply them
+    // to the histograms and get the total rows:
+    // ---------------------------------------------------------
+
+    // Get all the key preds for the key columns up to the first
+    // key column with no key preds
+    ValueIdSet singleSubsetPrefixPredicates;
+    NABoolean conflict = FALSE;
+    for ( CollIndex i = 0; i <= singleSubsetPrefixOrder; i++ )
+    {
+
+      const ValueIdSet *predsPtr = keyPredsByCol[i];
+      CMPASSERT( predsPtr != NULL ); // it must contain preds
+      singleSubsetPrefixPredicates.insert( *predsPtr );
+
+    } // for every key col in the sing. subset prefix
+
+    // Apply those key predicates that reference key columns
+    // before the first missing key to the histograms:
+    const SelectivityHint * selHint = CIDesc->getPrimaryTableDesc()->getSelectivityHint();
+    const CardinalityHint * cardHint = CIDesc->getPrimaryTableDesc()->getCardinalityHint();
+
+    RelExpr * dummyExpr = new (STMTHEAP) RelExpr(ITM_FIRST_ITEM_OP,
+				                 NULL,
+				                 NULL,
+				                 STMTHEAP);
+
+    histograms.applyPredicates( singleSubsetPrefixPredicates, *dummyExpr, selHint, cardHint);
+
+  } // if there are key predicates
+
+  // If there is no key predicates, a full table scan will be generated.
+  // Otherwise, key predicates will be applied to the histograms.
+  // Now, compute the number of rows after key preds are applied,
+  // and accounting for asynchronous parallelism:
+  CostScalar numRowsToScan =
+    ((histograms.getRowCount()/activePartitions).getCeiling()).minCsOne();
+
+  return numRowsToScan;
+}   // CostMethodHbaseUpdateOrDelete::numRowsToScanWhenAllKeyColumnsHaveHistograms()
+#pragma warn(262)  // warning elimination
+
+// -----------------------------------------------------------------------
+// CostMethodHbaseUpdateOrDelete::computeIOCostsForCursorOperation().
+// -----------------------------------------------------------------------
+void CostMethodHbaseUpdateOrDelete::computeIOCostsForCursorOperation(
+  CostScalar & randomIOs,        // out
+  CostScalar & sequentialIOs,    // out
+  const IndexDesc * CIDesc,
+  const CostScalar & numRowsToScan,
+  NABoolean probesInOrder
+  ) const
+{
+  const CostScalar indexLevels = CIDesc->getIndexLevels();
```

Comment:
IndexLevels is the number of index levels in a randomly selected HFile, excluding the data blocks.  See Java method FixedFileTrailer::getNumDataIndexLevels(). The JNI code for it is getHbaseTableInfo().